### PR TITLE
GPII-3397- Enable log exclusion for the CouchDB HTTP checks

### PR DESCRIPTION
**Pull request by amatas**
The log severity configured in the CouchDB containers is `info`, and this causes that the CouchDB logs every request from the health checks:
```
[notice] 2018-10-11T18:11:14.179590Z couchdb@couchdb-couchdb-1.couchdb-couchdb.gpii.svc.cluster.local <0.17722.0> 66cbd57ff5 localhost:5984 127.0.0.1 a99c4851ccfb586fa58727bcbb869a30 GET /_up 200 ok 2
```
This PR decreases the log level to `warn`.
http://docs.couchdb.org/en/stable/config/logging.html#log/level
which is less noisy.

### CODE REVIEW

File: gcp/modules/couchdb/values.yaml [hunk 1]
```diff
@@ -27,6 +27,8 @@ couchdbConfig:
     bind_address: any
     # http://docs.couchdb.org/en/stable/config/auth.html#authentication-configuration
     require_valid_user: true
+  log:
+    level: warn
```
Comment (natarajaya) [inline gcp/modules/couchdb/values.yaml]: Does this mean that we're not going to see actual HTTP requests to CouchDB anymore in logs? If it is the case, I think this is undesirable.

Comment (amatas) [reply gcp/modules/couchdb/values.yaml]: I understand, by the documentation, that we would only remove the 2XX requests logs, other requests as 5XX will be logged. http://docs.couchdb.org/en/stable/config/logging.html#log/level

Do you think that we should save ALL the HTTP requests? I think that those logs would demand more resources than the real value of that information.

**Comment by stepanstipl** [comment]
Looks good. Personally I would decrease the log level and test if we can increase it without disrupting ops should we need to debug something.

**Comment by amatas** [comment]
The logs that spam the stackdriver has the level of [notice], `warn` is the immediate next level.

**Comment by natarajaya** [comment]
Having full request logs for DB layer is essential for audit purposes. I think we should keep current log level, and if those health checks are really noisy, configure exclusion instead.

File: gcp/modules/couchdb/values.yaml (hunk 1 above)
Comment (natarajaya) [reply gcp/modules/couchdb/values.yaml]: Replied in thread.

**Comment by mrtyler** [comment]
I don't think it's correct to change the log level *or* to exclude this log message. My understanding is the cause of this message is a misconfigured liveness check. If so, I'd say the solution is to fix the liveness check. Silencing a log that revealed a real error seems counterproductive. Am I missing anything?

**Comment by amatas** [comment]
Let me summarize what we have.

If you go to your own dev cluster, at the log viewer and enter this filter (change your user in the project name):
```
resource.labels.cluster_name="k8s-cluster"
logName="projects/gpii-gcp-dev-<your_user>/logs/stderr"
resource.labels.container_name="couchdb"
```
you will see a lot of messages like this:
```
{
 insertId: "10crbmbg3hsmbrg"  
 logName: "projects/gpii-gcp-dev-alfredo/logs/stderr"  
 metadata: {…}  
 receiveTimestamp: "2018-10-16T17:51:51.412688468Z"  
 resource: {…}  
 severity: "ERROR"  
 textPayload: "[notice] 2018-10-16T17:51:47.560829Z couchdb@couchdb-couchdb-0.couchdb-couchdb.gpii.svc.cluster.local <0.31436.2> a0b399088c localhost:5984 127.0.0.1 629a540acb21d93a0f242baf2498b5d9 GET /_up 200 ok 2
"  
 timestamp: "2018-10-16T17:51:47Z"  
}
```
It you see at the end of the payload, it is a common request that returns a 200. It's an OK for the liveness and the readyness.

The problem here, IMHO, is that the CouchDB logs [every request to stderr](http://docs.couchdb.org/en/stable/config/logging.html#log/writer) and the [Stackdriver understands that the severity is "ERROR"](https://cloud.google.com/kubernetes-engine/docs/how-to/logging#best_practices) for all the logs that the CouchDB throws, making very difficult to select only a type of log later on. So we have that all the logs of the CouchDB are marked as ERROR from the same source, without any other tags that help us to discriminate between a "real error" and a "200 request".

I feel that I'm out of ideas. These are my proposes:
1. Leave as it is now. We will have a lot of 200 messages, but we would also collect the real errors.
1. Increase the [severity of the logs in the CouchDB](http://docs.couchdb.org/en/stable/config/logging.html#log/level). We will get rid of the `200 ok` logs and we would collect only other kind of messages.
1. Set an exclusion in Stackdriver. But since we don't have more fields to discriminate from one type of errors from others this could be hard to find out. The most accurate filter that I tried is:
```
resource.labels.cluster_name="k8s-cluster"
logName="projects/gpii-gcp-dev-<your_user>/logs/stderr"
resource.labels.container_name="couchdb"
```

**Comment by natarajaya** [comment]
> The problem here, IMHO, is that the CouchDB logs every request to stderr

@amatas I think this can be solved by configuring CouchDB to write logs into file:
http://docs.couchdb.org/en/stable/config/logging.html#log/writer
And then using sidecar container to stream them, similarly to this example:
https://kubernetes.io/docs/concepts/cluster-administration/logging/#streaming-sidecar-container

Probably worth trying. It would've been nice if we can have CouchDB logs with adequate severity. Not sure if it is required to configure health check requests exclusion though. This filter works for me (I know exact matches are faster, but since we don't have this option in our case):
```
resource.type="k8s_container"
resource.labels.container_name="couchdb"
textPayload : "GET / 200 ok 2"
```

**Comment by mrtyler** [comment]
@stepanstipl 
Ah, I mistook the log message in this PR for the log message in the associated ticket -- the subtly but significantly different `undefined GET / 200 ok 1`.

@amatas 
Thank you for the summary and explanation. I agree with @natarajaya that it seems the root cause is the disagreement between CouchDB and Stackdriver about what it means to write a message to stderr. However, I think a solution involving a sidecar is not worth the effort at this time.

An exclusion is cheap and addresses my main concern (keeping clutter out of the logs to make debugging easier) and my secondary concern (minimizing the amount and therefore cost of the logs we archive). I think an exclusion is the right move.

Sergey's filter looks good. I would add a space before `GET` because I'm always paranoid about false positives.

One thing that is unfortunate about this solution is the high coupling / low cohesion between this filter and the liveness check in the couchdb chart (especially with a lack of automated testing to tell us when we've changed one without changing the other). I can think of some solutions (make liveness check URL a rake variable and pass it to various terraform modules; couchdb module includes its own instance of gcp-stackdriver-exclude to manage this exclusion) but I don't think any of them are worth the trouble now.

**Comment by stepanstipl** [comment]
@amatas thanks for the explanation. Can't we make CouchDB write logs to stdout? Otherwise this will be an issue with every single event - everything will be logged  with `severity: "ERROR"`, right?

I would prefer to fix this at the source (as opposed to fixing by exclusion). Also this will be an issue with log exports, the exclusions do not apply there, right?

**Comment by amatas** [comment]
> @amatas thanks for the explanation. Can't we make CouchDB write logs to stdout? Otherwise this will be an issue with every single event - everything will be logged with `severity: "ERROR"`, right?

I guess that we will have the same amount of logs but with the severity field with other value. That wouldn't decrease the amount of logs received.

> I would prefer to fix this at the source (as opposed to fixing by exclusion). Also this will be an issue with log exports, the exclusions do not apply there, right?

I agree with you @stepanstipl , but the CouchDB doesn't allow to set which logs goes to what severity. Perhaps putting an application in the middle that digests the logs and send them properly to the stackdriver would help. But it also adds some complexity. As @mrtyler and @natarajaya said before, perhaps we can improve this in the future.

**Comment by mrtyler** [comment]
LGTM

File: gcp/modules/gcp-stackdriver-exclusion/main.tf [hunk 2]
```diff
@@ -19,6 +19,10 @@ variable "exclusions" {
     # 0") from a string that happens to start with 200 ("Setting calico-typha
     # requests["cpu"] = 200m").
     "kubelet-stats-summary-200" = "resource.type=k8s_node AND logName=projects/__var.project_id__/logs/kubelet AND jsonPayload.MESSAGE:GET /stats/summary AND jsonPayload.MESSAGE:\" 200 \""
+
+    "couchdb-liveness-check-200" = "resource.type=k8s_container AND resource.labels.container_name=couchdb AND textPayload:\" GET / 200 ok \""
+
+    "couchdb-readyness-check-200" = "resource.type=k8s_container AND resource.labels.container_name=couchdb AND textPayload:\" GET /_up 200 ok \""
```
Comment (natarajaya) [inline gcp/modules/gcp-stackdriver-exclusion/main.tf]: It is readiness.

**Comment by natarajaya** [comment]
LGTM!

Can you please update this PR's title so we know it is about logging exclusions and not severity?

**Pull request merged**